eth/catalyst/api.go

```diff
@@ -361,14 +420,63 @@ func (api *ConsensusAPI) delayPayloadImport(block *types.Block) (beacon.PayloadS
 	return beacon.PayloadStatusV1{Status: beacon.ACCEPTED}, nil
 }
 
+// setInvalidAncestor is a callback for the downloader to notify us if a bad block
+// is encountered during the async sync.
+func (api *ConsensusAPI) setInvalidAncestor(invalid *types.Header, origin *types.Header) {
+	api.invalidLock.Lock()
+	defer api.invalidLock.Unlock()
+
+	api.invalidTipsets[origin.Hash()] = invalid
+}
+
+// checkInvalidAncestor checks whether the specified chain end links to a known
+// bad ancestor. If yes, it constructs the payload failure response to return.
+func (api *ConsensusAPI) checkInvalidAncestor(check common.Hash, head common.Hash) *beacon.PayloadStatusV1 {
+	api.invalidLock.Lock()
+	defer api.invalidLock.Unlock()
+
+	// If the hash to check is unknown, return valid
+	invalid, ok := api.invalidTipsets[check]
+	if !ok {
+		return nil
+	}
+	// If the bad hash was hit too many times, evict it and try to reprocess in
+	// the hopes that we have a data race that we can exit out of.
+	badHash := invalid.Hash()
+
+	api.invalidBlocksHits[badHash]++
+	if api.invalidBlocksHits[badHash] >= invalidBlockHitEviction {
```

Comment:
It's problematic to me. 

If we try to re-process a block and eventually it's adopted this time, we only delete "hit marker", but leave "invalidTipSet" unchanged with all descendants still marked as BAD.

Let's imagine the scenario:

- We have BAD BLOCK X, its descendant X+1 are also marked in the "invalidTipSet"
- Process X, Block X gets adopted somehow, and `invalidBlocksHits[X]` is reset to 0
- Process X+1, `invalidTipsets[X+1]` is still existent and `invalidBlocksHits[X]` is 0
- X+1 will be rejected

Reply:
I'm kind of wondering how we expect the CL to operate here. Considering that geth deems `Bn` to be invalid, would the CL

a) Try to import it over and over `Bn, Bn, Bn, Bn....`, or
b) Try to import the tips after it: `Bn, Bn+1, Bn+2...`. 

This PR lgtm if we expect scenario a), but probably won't work great if we expect scenario b). 

?

Reply:
I think both of them are expected. Different CL's have different strategies regarding retrying invalid blocks/syncing etc

Reply:
> I'm kind of wondering how we expect the CL to operate here. Considering that geth deems `Bn` to be invalid, would the CL
> 
> a) Try to import it over and over `Bn, Bn, Bn, Bn....`, or b) Try to import the tips after it: `Bn, Bn+1, Bn+2...`.
> 
> This PR lgtm if we expect scenario a), but probably won't work great if we expect scenario b).
> 
> ?

Both scenarios are handled. When a NewPayload is received, we check the block against the set of invalid blocks. If it is present, we reject if (scenario A). If the block is not present, we try to retrieve the parent state (to process on top) and if that's missing, we check the parent's hash for bad block-ness. If it is, we reject the payload (scenario B).

Reply:
> It's problematic to me.
> 
> If we try to re-process a block and eventually it's adopted this time, we only delete "hit marker", but leave "invalidTipSet" unchanged with all descendants still marked as BAD.
> 
> Let's imagine the scenario:
> 
> * We have BAD BLOCK X, its descendant X+1 are also marked in the "invalidTipSet"
> * Process X, Block X gets adopted somehow, and `invalidBlocksHits[X]` is reset to 0
> * Process X+1, `invalidTipsets[X+1]` is still existent and `invalidBlocksHits[X]` is 0
> * X+1 will be rejected

Good catch. We need to clean up all the future blocks referencing the same bad block.